server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java

```diff
@@ -35,6 +36,61 @@
  * <p>
  * Similar to {@link ListenableActionFuture} and {@link ListenableFuture} except for its handling of exceptions: if this listener is
  * completed exceptionally then the exception is passed to subscribed listeners without modification.
+ * <p>
+ * Often this will be used to chain together a sequence of async actions, similarly to {@link CompletionStage} (without the
+ * {@code catch (Throwable t)}), such as in the following example:
+ * <pre>
+ * private void exampleAsyncMethod(String request, List&lt;Long> items, ActionListener&lt;Boolean> finalListener) {
+ *     SubscribableListener
+ *
+ *         // Start the chain and run the first step by creating a SubscribableListener using newForked():
+ *         .&lt;String>newForked(l -> firstAsyncStep(request, l))
+ *
+ *         // Run a second step when the first step completes using andThen(); if the first step fails then the exception falls through to
+ *         // the end without executing the intervening steps.
+ *         .&lt;Integer>andThen((l, firstStepResult) -> secondAsyncStep(request, firstStepResult, l))
```

Comment:
Is it always necessary to specify the type in the `.<T>newForked` and `<T>andThen` operations? I wonder why java cannot find them automatically.

Reply:
Not always, but yes they're often necessary unless you put the types somewhere else. The type checker doesn't do a great job with `Doubly<Nested<Generics>>` it seems. That's another good reason for having this example in the test suite, so that we can see we're not missing anything.